### ISSUE TITLE
Vision Pro Dev strap streaming

### ISSUE THREAD

**Issue by dougshean** [alvr-org/alvr-visionos#106]
Hi does alvr support streaming through usb-c connection?  Has anybody tried it?  Thanks

**Comment by s-triant** [comment]
> You need to connect the dev-strap with USB-C to a Mac, connect the Mac with Ethernet to a PC, and then run some ifconfig commands on the Mac to forward the traffic to the PC, after which the PC will recognize it and you can manually enter the wired IP of the Vision Pro into the ALVR server. The last time I used ALVR on the Vision Pro was about 2 months ago, so there might be a different way now.

Have you actually done this and it worked? I am likely to purchase the strap if it actually works.

**Comment by AlexTech01** [comment]
> Have you actually done this and it worked? I am likely to purchase the strap if it actually works.

Yup, there are some ifconfig commands you need on the Mac though. I also found another method which involves setting up a Linux VM on the Windows PC, configuring USB sharing so the dev-strap's USB signal gets sent to the VM, and then configuring networking on the VM so the USB signal from the dev-strap gets sent back to Windows. For whatever reason this will make Windows identify it correctly and it'll be able to actually establish a link

**Comment by shinyquagsire23** [comment]
@AlexTech01 ok that's both genius and also extremely cursed, wonder if someone has a Linux microVM I could package up specifically for that purpose lmao

**Comment by AlexTech01** [comment]
> @AlexTech01 ok that's both genius and also extremely cursed, wonder if someone has a Linux microVM I could package up specifically for that purpose lmao

@shinyquagsire23 I don't know much about microVMs, but here's a VMware Workstation 17 Pro VM with some changes made so it'll bridge the AVP to the host: https://drive.google.com/file/d/1xdj_3pkgthv6utfg24WIXx821UsiHfZy/view?usp=sharing

To import in VMware Workstation 17 Pro:
1. Extract the downloaded zip file
2. Open VMware Workstation 17 Pro
3. Click File -> Open (or Ctrl+O)
4. Select the OVF file
5. Give your VM a name and location
6. VMware should automatically locate the VMDK disk file when you try to import the OVF, so you don't have to import it manually
7. The VM will be created and you can start it

Make sure your settings in VMware are set to either ask when a USB is connected or to automatically connect the USB to the VM (asking when a USB is connected is preferred.) You can change this setting by opening Edit -> Preferences (Ctrl+P), going to the USB tab, and changing the setting.

Here's a list of the changes I've made to this Ubuntu 24.04 Server minimized instance:
1. open-vm-tools and nano are installed via apt
2. systemd-networkd-wait-online.service is masked in order to allow the system to boot with no internet connection
3. There is an "/etc/udev/rules.d/99-usb-net-bridge.rules" file to handle plugging in/removing the AVP from the VM
4. There is an "/etc/netplan/99-bridge-config.yaml" file to make the network bridge (**note**: you might have to change the 2 mentions of "ens37" in here to the host-only interface your VM sees, you can find this with "ip addr")
5. There are 2 scripts in "/var/ncm/scripts" to handle plugging in and removing USB devices, this is what the udev rules execute
6. There is a folder "/var/ncm/config" that can be used to change some basic config things, the existence of a `service_enabled` file makes the script work and the existence of `logging_enabled` makes it log USB connections/disconnections to /var/log/ncm.log. By default, the service_enabled file exists, but not the logging_enabled file.
7. There is an "ncmctl" bash script in /usr/local/bin that can be used to disable/enable the service and logging if you don't want to manually create files. Use `ncmctl enable`, `ncmctl disable`, `ncmctl status`, etc. to manage the service, and `ncmctl logging enable`, `ncmctl logging disable`, `ncmctl logging status`, to manage logging. Anything other than the status command will require sudo.

The default username/password is `ncmbridge`/`ncmbridge` and that can be used to login to the VM to look at these scripts and make changes. The ncmbridge user also has sudo access.

**Comment by PedroM4** [comment]
@AlexTech01 

would you be so kind as to measure the max bandwith the AVP dev strap is capable of supplying?
You can easily do this using iperf3 running as server on the PC  ( iperf3 -s ) and installing the iperf3 client on AVP ( it's available from the app store ).

**Comment by shinyquagsire23** [comment]
@PedroM4 It's 240Mbps, basically identical to WiFi 5GHz

**Comment by s-triant** [comment]
> @PedroM4 It's 240Mbps, basically identical to WiFi 5GHz

If I am not mistaken, AVP is capable of 1,200Mbps over Wi-Fi, or up to five times faster than through the dev strap.

**Comment by shinyquagsire23** [comment]
yeah in practice it seems to bump up against latency issues around there, not sure why (maybe it's a decoder bandwidth limitation idk)

**Comment by PedroM4** [comment]
ok, so not that bad. I'm getting 960mbps average on the wifi and can push the decoder to about 550mbps at least that is what ALVR bandwidth plot says, but there is a catch, I move my head and the bandwidth may drop to 300mps or so, and that's really not nice. 
I think stable 240mbps may be better than an wifi connection that is so unstable but I would have to try it. 
I'll wait for the dev strap to be available in my region.

Anyway, I was reading a bit about it today and it seems that the speed limit is a software imposed thing, so maybe in the future apple changes that or a third party vendor appears with a wired solution.

**Comment by BitMangler** [comment]
Using a VMware VM for 'ANY Part' of the VR streaming path from a PC to the AVP is a remakably STUPID thing to do.   The task/thread scheduling of a VM in ANY modern OS (like macOS or Win-10/11) will add SO MUCH latency to that 
"PC to VR Headset" path that it will ruin most VR experiences, if it even works at all.   I have not tried it, but I was a pioneer in bringing Hypervisors (VM techno) to PC's...starting back in the mid 80's.   So I know what I'm talking about in terms of VM's getting  unreal latency for communications tasks.   Don't believe me, just try to watch an HD video inside a VM...see how well that works out for you.   You'll get so much hiccup that you'll quickly lose interest in watching or listening.

**Comment by shinyquagsire23** [comment]
@BitMangler yeah I tried it briefly and opted for an RPi, but it seems visionOS 26 at least unlocked proper USB host support, so now you can just use a USB C ethernet dongle and it's fine

**Comment by BitMangler** [comment]
@shinyquagsire23 Don't get me wrong, Hypervisor technology (VMs) are useful for many things, but when real time is required (as so many comms tasks require), VMs just are not the way to fly your data.   And, that's even if you can strip the base operating system down to as minimum possible...deleting drivers, back ground processes, etc.   Where you try to the the System running NOTHING but your VM.   But, for most comms tasks, you'll just never get there.   The base HLOS will just always have enough interrupts and higher priority tasks, that whatever you have running in your VM will croak like frog ;-)

I worked on VM/386, with help of people at Intel.  Per IBM's allowance, we ripped off the architecture of VM/370 Hypervisor from IBM Mainframes to run on John Crawford's new V86 mode.   The very first "virtualization" hardware to exist on a microprocessor (the 80386).   John Crawford was the architect of the 8036, and they were right across the street from Softguard where we created VM/386 (the very first hypervisor to run on a micro-computer).

**Comment by Xericore** [comment]
> You need to connect the dev-strap with USB-C to a Mac, connect the Mac with Ethernet to a PC, and then run some ifconfig commands on the Mac to forward the traffic to the PC, after which the PC will recognize it and you can manually enter the wired IP of the Vision Pro into the ALVR server. The last time I used ALVR on the Vision Pro was about 2 months ago, so there might be a different way now.

Do you happen to know (roughly) which ifconfig commands you used? I'm trying to replicate the same, but I can't find the Vision Pro in the ALVR Dashboard. I've tried bridging the RJ45 Lan interface with the Vision Pro USB interface via these commands:

`sudo ifconfig bridge create
sudo ifconfig bridge0 addm en0 addm en1 up
sudo ifconfig bridge0 inet 192.168.X.X netmask 255.255.255.0`

But I've had no success with it.

**Comment by newbishme** [comment]
> > You need to connect the dev-strap with USB-C to a Mac, connect the Mac with Ethernet to a PC, and then run some ifconfig commands on the Mac to forward the traffic to the PC, after which the PC will recognize it and you can manually enter the wired IP of the Vision Pro into the ALVR server. The last time I used ALVR on the Vision Pro was about 2 months ago, so there might be a different way now.
> 
> Do you happen to know (roughly) which ifconfig commands you used? I'm trying to replicate the same, but I can't find the Vision Pro in the ALVR Dashboard. I've tried bridging the RJ45 Lan interface with the Vision Pro USB interface via these commands:
> 
> `sudo ifconfig bridge create` `sudo ifconfig bridge0 addm en0 addm en1 up` `sudo ifconfig bridge0 inet 192.168.X.X netmask 255.255.255.0`
> 
> But I've had no success with it.

this works, give it a try: https://gist.github.com/rebelancap/ac3ab6c82c7ac6c388969b6586fd567c 

```
PRIMARY_INTERFACE=$(route -n get default 2>/dev/null | grep interface | awk '{print $2}')
NETWORK_INFO=$(ifconfig "$PRIMARY_INTERFACE" | grep "inet " | awk '{print $2, $4}')
IP_ADDR=$(echo $NETWORK_INFO | awk '{print $1}')
NETWORK_BASE=$(echo $IP_ADDR | cut -d. -f1-3)
BRIDGE_IP="${NETWORK_BASE}.200"

sudo sysctl -w net.inet.ip.forwarding=1 > /dev/null
sudo ifconfig bridge0 inet $BRIDGE_IP netmask 255.255.255.0
sudo ifconfig bridge0 addm $PRIMARY_INTERFACE 2>/dev/null || true
sudo pfctl -e 2>/dev/null || true
echo "nat on $PRIMARY_INTERFACE inet from bridge0:network to any -> ($PRIMARY_INTERFACE)" | sudo pfctl -f - 2>/dev/null
```

**Comment by kwcnldnsldskcls** [comment]
Are there any news to this topic? Is there a more easy way now to use the strap with windows? for example only activating the wired connection in alvr on windows and connect via usb c? Would be super helpful, thank you! :)
